### PR TITLE
Purchase order registry

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR add an implementation of the purchase order registry and make some fixing.

- [x] Fix purchase order receipt
- [x] Add purchase order registry 
- [x] Add status for purchase order 
- [x] E2E test for the purchase order registry

### CODE REVIEW

File: client/src/js/services/SearchFilterFormatService.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+    /**
+     * @function formatFilterParameters
+     * @description filter and format properties of an object for being processed by formatFilterValues
+     * @param {object} element the object which contains all filters and their values. Ex. { reference: 'TPA1', is_cancelled: 1 }
+     * @param {boolean} WITH_NULL_VALUES a boolean flag for returning also properties which have null values 
+     */
+    function formatFilterParameters(element, WITH_NULL_VALUES) {
+        var out = {};
+        for (var i in element) {
```
Comment (jniles) [inline client/src/js/services/SearchFilterFormatService.js]: Can this be:
```js
var out = {};
angular.forEach(element, function (value, key) {
  if (WITH_NULL_VALUES || value) {
    out[key] = value;
  }
});
return out;
```
It will stop jshint from complaining about not using `Object.hasOwnProperty()` and has only one `if()` check.

File: client/src/js/services/SearchFilterFormatService.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+   * @function formatFilterValues
+   * @description identifier and display value
+   * @param {object} formatedFilters a returned value of formatFilterParameters
+   * @return {object} fomatedValues { identifiers: {}, display: {} }
+   */
+  function formatFilterValues(formatedFilters) {
+    var out = { identifiers: {}, display: {} };
+    for (var key in formatedFilters) {
+
+      if (!formatedFilters.hasOwnProperty(key)) { continue; }
```
Comment (jniles) [inline client/src/js/services/SearchFilterFormatService.js]: Using [angular.forEach()](https://docs.angularjs.org/api/ng/function/angular.forEach) will prevent the need for this type of code.

File: client/src/js/services/SearchFilterFormatService.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+      // get identifiers
+      out.identifiers[key] = typeof(formatedFilters[key]) === 'object' ?
```
Comment (jniles) [inline client/src/js/services/SearchFilterFormatService.js]: For clarity, it would be nice to have this check in a function:
```js
// takes in an object of filter and key and returns the identifier for the filter.
function getFormattedFilterIdentifier(formattedFilters, key) {
  var filter = formattedFilters[key];  // get the filter
  return angular.isObject(filter) ?  filter.uuid || filter.id || filter  : filter;
}

// .... later ....

out.identifiers[key] = getFormattedFilterIdentifier(formatedFilters, key);
```

File: client/src/js/services/SearchFilterFormatService.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  /**
+   * @function onRemoveFilter
+   * @description when remove a filter reload data
+   * @param {object} filters {identifiers: ..., display: ...}
+   * @param {function} reload A reload function 
+   */
+  function onRemoveFilter(key, filters, reload) {
+    if ((filters && !filters.display && !filters.identifiers) || (!filters) || !reload) { return; }
```
Comment (jniles) [inline client/src/js/services/SearchFilterFormatService.js]: This is super hard to parse.  Can you find a way to clean this up a bit?  Maybe with variable names?

```js
var noFilters = !filters;
var invalidFilterFormat = (filters && !filters.display && !filters.identifiers);
var noReload = !reload:
if (invalidFilterFormat || noFilters || noReload) { return; }
```

File: client/src/js/services/SearchFilterFormatService.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    
+    if (key === 'dateFrom' ||  key === 'dateTo') {
```
Comment (jniles) [inline client/src/js/services/SearchFilterFormatService.js]: Why do we remove the entire date range if one of the dates is deleted?

File: client/src/partials/purchases/list/list.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+      action: search, color: 'btn-default'
+    }
+  ];
+
+  var columnDefs  = [
+    { 
+        field : 'reference', 
+        displayName : 'FORM.LABELS.REFERENCE', 
+        headerCellFilter : 'translate', 
+        aggregationType: uiGridConstants.aggregationTypes.count },
```
Comment (jniles) [inline client/src/partials/purchases/list/list.js]: Since we don't have proper i18n for the grid, I encourage you to use `aggregationHideLabel: true` on all the columns that have aggregations.

File: client/src/partials/purchases/list/list.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  // add purchase order 
+  function addPurchaseOrder() {
+    $state.go('/purchases/create');
+  }
+
+  // get document 
+  function getDocument(uuid) {
+    Receipts.purchase(uuid, true);
```
Comment (jniles) [inline client/src/partials/purchases/list/list.js]: Should this actually be `true`?   Won't that show the created text?

File: client/src/partials/purchases/modals/search.tmpl.html [hunk 8]
```diff
@@ -0,0 +1,98 @@
+<form
+  name="ModalForm"
+  ng-submit="$ctrl.submit(ModalForm)"
+  data-modal="purchase-search"
+  bh-form-defaults
+  novalidate>
+
+  <div class="modal-header">
+    <ol class="headercrumb">
+      <li class="static">{{ "PURCHASES.ORDER" | translate }}</li>
```
Comment (jniles) [inline client/src/partials/purchases/modals/search.tmpl.html]: Remember, the translate directive is more efficient than the translate filter.

File: client/src/partials/purchases/modals/search.modal.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  // global methods
+  vm.validate = validate;
+  vm.cancel = Instance.close;
+  vm.submit = submit;
+
+  // init 
+  init();
+
+  // load users 
+  Users.read(null, { detailled: 1 })
```
Comment (jniles) [inline client/src/partials/purchases/modals/search.modal.js]: Do you need detailed information about the user?  I think you can get away with `Users.read()`.

File: client/src/partials/purchases/modals/status.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    // expose to view 
+    vm.close = Instance.close;
+    vm.submit = submit;
+
+    // submit the choice 
+    function submit(status) {
+        var data;
+
+        switch(status) {
```
Comment (jniles) [inline client/src/partials/purchases/modals/status.js]: Can we just do:
```js
var isConfirmed = status === 'confirmed' ? 1 : 0;
var data = { is_confirmed : isConfirmed, is_cancelled: vm.purchase.is_cancled };
updateOrder(vm.purchase.uuid, data);
```
instead of these ~12 lines?

File: client/src/partials/purchases/modals/status.tmpl.html [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+          <span class="label label-primary" translate>PURCHASES.STATUS.CONFIRMED</span>
+          <p translate>PURCHASES.STATUS.CONFIRMED_DESC</p>
+      </label>
+  </div>
+
+  <div class="form-group">
+      <label class="text-action">
+          <input type="radio" name="status" ng-model="$ctrl.status" value="not_confirmed"> 
+          <span class="label label-default" translate>PURCHASES.STATUS.WAITING_CONFIRMATION</span>
+          <p translate>PURCHASES.STATUS.WAITING_CONFIRMATION_DESC</p>
```
Comment (jniles) [inline client/src/partials/purchases/modals/status.tmpl.html:29]: :+1: Nice use of the translate directive.

File: client/src/partials/purchases/list/list.js [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+      .then(function (filters) {
+        if (!filters) { return; }
+        reload(filters);
+      })
+      .catch(Notify.handleError);
+  }
+
+  // add purchase order 
+  function addPurchaseOrder() {
+    $state.go('/purchases/create');
```
Comment (jniles) [inline client/src/partials/purchases/list/list.js]: This should be `$state.go(purchasesCreate);`

File: client/src/partials/purchases/modals/search.modal.js [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+  .then(function (users) {
+      vm.users = users;
+  })
+  .catch(Notify.handleError);
+
+  // load suppliers 
+  Suppliers.read(null, { detailled: 1 })
```
Comment (jniles) [inline client/src/partials/purchases/modals/search.modal.js]: The `detailled` option doesn't exist for suppliers.

File: client/src/partials/purchases/templates/cellStatus.tmpl.html [hunk 14]
```diff
@@ -0,0 +1,17 @@
+<div class="ui-grid-cell-contents">
+    <div ng-if="row.entity.is_confirmed === 1 && row.entity.is_cancelled === 0 && row.entity.is_received === 0" class="label label-primary" translate>
```
Comment (jniles) [inline client/src/partials/purchases/templates/cellStatus.tmpl.html]: It seems to me that you could reduce the number of checks here by nesting them so that angular doesn't execute all these checks on every change.

File: server/controllers/inventory/reports/purchases.receipt.handlebars [hunk 15]
```diff
@@ -1,40 +1,57 @@
-{{> head title="PATIENT_REG.PATIENT_REGISTRATIONS"}}
+{{> head title="PURCHASES.RECEIPT.TITLE"}}
```
Comment (jniles) [inline server/controllers/inventory/reports/purchases.receipt.handlebars:1]: :dancer:

**Review by jniles — COMMENTED** [review]
Just leaving my comments - I'll defer to @sfount for his review.

File: client/src/partials/purchases/list/list.js (hunk 6 above)
Comment (mbayopanda) [reply client/src/partials/purchases/list/list.js]: 👍

File: client/src/js/services/SearchFilterFormatService.js (hunk 4 above)
Comment (mbayopanda) [reply client/src/js/services/SearchFilterFormatService.js]: 👍

File: client/src/js/services/SearchFilterFormatService.js (hunk 1 above)
Comment (mbayopanda) [reply client/src/js/services/SearchFilterFormatService.js]: 🕺 👍

File: client/src/js/services/SearchFilterFormatService.js (hunk 3 above)
Comment (mbayopanda) [reply client/src/js/services/SearchFilterFormatService.js]: I did try this suggestion in a `angular.forEach` loop but it doesn't work, so i just create `getFormattedFilterIdentifier` function and i use it in a normal `for(...)` loop.

**Comment by jniles** [comment]
@mbayopanda, I propose to fork this PR, resolve the conflicts, and submit it for review by @sfount.  I think that the `SearchFilterFormatService.js` in this PR will help clean up the filters in all registries.  What do you think?

**Comment by mbayopanda** [comment]
@jniles that sound good. I will do it. Thanks

**Comment by jniles** [comment]
@mbayopanda, to be clear, I was proposing that I fork it and work on it. :)  Is that okay with you?

**Comment by mbayopanda** [comment]
There are 3 conflicts that i found and fixed them when i did merge it with
master.

I make a last push and you can fork it.

2017-02-14 4:49 GMT-08:00 Jonathan Niles <notifications@github.com>:

**Comment by mbayopanda** [comment]
I have just send the last push which fix the 3 conflicts.

There is no problem, you can tackle it.

Thanks for the help :+1

2017-02-14 9:58 GMT-08:00 Bruce Mbayo Panda <mbayopanda@gmail.com>:

File: client/src/js/app.js [hunk 16]
```diff
@@ -6,198 +6,190 @@ var bhima = angular.module('bhima', [
...
+        controller  : 'DashboardPatientController as PatientCtrl',
+      },
+    },
   })
   .state('exchange', {
-    abstract : true,
-    url : '/exchange',
-    templateUrl: 'partials/application/exchange.html'
+    abstract    : true,
+    url         : '/exchange',
```
Comment (jniles) [inline client/src/js/app.js:42]: :+1: Excellent adaptation to ESLint

File: client/src/js/app.js [hunk 17]
```diff
@@ -206,22 +198,20 @@ function translateConfig($translateProvider) {
...
 function startupConfig($rootScope, $state, $uibModalStack, SessionService, amMoment, Notify, $location) {
-
   var loginStateRegexp = /#\/login$/;
   var rootStateRegexp = /#\/$|\/$|#$/;
 
 
   // make sure the user is logged in and allowed to access states when
   // navigating by URL.  This is pure an authentication issue.
-  $rootScope.$on('$locationChangeStart', function (event, next) {
+  $rootScope.$on('$locationChangeStart', (event, next) => {
```
Comment (jniles) [inline client/src/js/app.js:214]: Unfortunately, we cannot use arrow functions in the browser.  They break old versions of chrome and break our gulp build :-1:

**Review by jniles — COMMENTED** [review]
This is excellent.  I'll fork this and get it in tomorrow!
